fix(es/preset-env): Consider `browserslist` config if `env.target` is not configured

egfx-notifications opened the pull request:
<!-- Note: CI script will automatically rebase your PR so please do not rebase unless required -->

<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**Description:**

The `env.path` option in `.swcrc` was originally implemented for `browserslist`, but implementation was (unintentionally?) removed during the migration to `browserslist-rs` (#2845).
Searching the current directory for browserslist configuration was also dropped in the process.

This PR reimplements searching for browserslist configuration using the lookup functionality provided by `browserslist-rs`

**BREAKING CHANGE:**

<!--
If this PR introduces a breaking change, it must contain a notice for it to be included in the CHANGELOG. Add description or remove entirely if not breaking.

You may need to update `MIGRATION.md` for the breaking changes.
-->
This may be considered a breaking change since the previous (broken) behavior for unconfigured `env.targets` was to return a `BrowserData` struct with all `Option`s set to `None`. Now the result of a browserslist `defaults` query is returned instead. This probably was the original behavior with `browserslist`, though.
*Update:* Thinking about it again, `targets_to_versions()` is part of `preset_env_base`s public API so we should probably consider this a breaking change anyway.

**Things to consider:**
- only works for non wasm targets as `browserslist-rs` does not provide configuration loading for wasm32 targets (which is probably to be expected that we won't check environment variables or search a filesystem in that case)
- `feat(preset-env): Let browserslist-rs handle default config path` changes the default for an unconfigured `env.path` from an empty string for wasm targets / current dir for other targets to a `None` value. `browserslist-rs` already looks up the current dir if `path` is `None` and it felt better to let it handle this case itself
  The commit which migrated from `browserslist` to `browserslist-rs` mentioned fixing the `default_path`, but does not mention what was broken and I also did not find any other usage of the `path` property. If the previous implementation was intentional, just ignore that last commit in the PR, the first two sufficiently address the bug.

**Related issue:**
Fixes  #3365
Implements #3085

egfx-notifications committed:
fix(preset-env): Load browserslist config if no swc target configured

egfx-notifications committed:
fix(preset-env): Pass through path from swc env config to browserslist

Defaults to empty PathBuf for wasm32 target, but does not seem to be
used anyway

egfx-notifications committed:
feat(preset-env): Let browserslist-rs handle default config path

egfx-notifications committed:
fix(preset-env): Ignore env.path for swc_css_prefixer

Maybe reconsider this later if this is functionality we want

egfx-notifications commented:
It seems some of the tests fail due to them expecting a specific prerecorded results, but this result has now changed due to the previous default without a target being `BrowserData` with only `None` values and now it is `BrowserData` according to browserslist-rs `defaults` query, so some stuff that browser now understand natively is not transpiled.

Seems forcing a browserslist query when target is none is not the best approach. I won't try any fixes before hearing your feedback on how to handle this. Should we make the `env.path` option mandatory if a browserslist config lookup is desired?

kdy1 commented on crates/preset_env_base/src/query.rs:
```diff
@@ -96,8 +96,31 @@ impl Query {
+                        .unwrap_or_else(|| panic!("Invalid path \"{:?}\"", path)),
+                );
+            }
+            let distribs = browserslist::execute(&browserslist_opts)
```
Does this work with configs that execute javascript? cc @g-plane

g-plane replied:
No, it doesn't execute any code.

kdy1 reviewed (commented):
Can you rebase?

jodaka commented:
Nice to see some action here! Thanks @egfx-notifications

egfx-notifications commented:
I think I addressed all of the changes between my original commits and now. Unfortunately I can't run `cargo test` locally right now as my GLIBC version is too old. Let's see what CI says. I may set up a newer Ubuntu VM tomorrow if necessary.

kdy1 commented:
No worries, I‘ll update the tests if needed

kdy1 committed:
Create tidy-pots-report.md

kdy1 commented on .changeset/tidy-pots-report.md:
```diff
@@ -0,0 +1,5 @@
+---
+preset_env_base: major
```
As this is a breaking change of swc_core, I'll merge it later.

kdy1 commented on crates/preset_env_base/src/query.rs:
```diff
@@ -96,8 +96,35 @@ impl Query {
+                    .clone()
+                    .into_os_string()
+                    .into_string()
+                    .unwrap_or_else(|_| panic!("Invalid path \"{:?}\"", path))
```
I think we should not `panic!()` here. Can you return an instance of `anyhow::Error` instead?

kdy1 commented on crates/preset_env_base/src/query.rs:
```diff
@@ -96,8 +96,35 @@ impl Query {
+                .with_context(|| "failed to resolve browserslist query from browserslist config")?;
+
+            let versions =
+                BrowserData::parse_versions(distribs).expect("failed to parse browser version");
```
Same here. We don't have to panic.

egfx-notifications committed:
fix(preset-env): Prefer anyhow:Error over panic!

kdy1 reviewed (dismissed):
Thank you, and sorry for a (very) late review!

kdy1 reviewed (commented):
I missed it while reviewing, but while trying to update tests to merge this PR, I found that this PR makes the `swc` crate ignore `jsc.target`.
Could you fix it?

You can run `UPDATE=1 cargo test --test projects --test tsc --test error_msg` from `./crates/swc` to update test snapshots.

egfx-notifications commented:
> I missed it while reviewing, but while trying to update tests to merge this PR, I found that this PR makes the `swc` crate ignore `jsc.target`. Could you fix it?

So, I looked into this and think I need a little more context. The only test failures I could find in the fixture tests were tests which explicitly **didn't** specify `jsc.target`. Those test cases have neither `env.target` nor `jsc.target` defined and as I mentioned the default if specifying no target is now the result of a browserslist `defaults` query, thus making this PR a breaking change. So those are test failures that were expected to happen until the fixtures are updated.
Note that the shopify tests 003-005 also fall into this "defaults" change since their `.swcrc` is wrong, they specify 
```json
{
    "jsc": {
        "parser": {
            "syntax": "ecmascript",
            "jsx": true
        }
    },
    "module": {
        "type": "commonjs"
    },
    "env": {
        "chrome": "50"
    }
}
```
when it should have been
```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "jsx": true
    }
  },
  "module": {
    "type": "commonjs"
  },
  "env": {
    "targets": {
      "chrome": "50"
    }
  }
}
```
So those fixtures weren't using that targeted chrome version from the start.

As far as I can tell `jsc.target` should still work as intended if specified.

kdy1 commented:
Oh, I think then it's fine and we should specify `jsc.target: es5` for those tests I think

egfx-notifications committed:
fix(preset-env): Make fixture tests independent of browserslist defaults query

The previous browserslist implementation without autodiscovery of .browserslistrc defaulted to applying all possible transformations if no target was configured. With autodiscovery there will always be a default target defined by browserslist itself. By adding explicit .browserslistrc files with empty targets we recreate the original default for the tests so the fixtures do not need to be rerecorded whenever browserslist updates their default.

egfx-notifications committed:
test(preset-env): Add test for env.path configuration option

If env.path would be ignored we would receive var instead of const in the output

egfx-notifications commented:
So, since `jsc.target` and `env` can't be used together and I think the `env` configurations in those test cases may be relevant for the case, I tried another route to solve this. I put `.browserslistrc` files without any content into the crate directories of the failing tests. This way it is picked up by autodiscovery if the test doesn't specify `jsc.target` or `env.targets`, but an empty query results in applying all possible transformations and thus emulates the previous default. This also means we don't have to regenerate all fixtures everytime `browserslist-rs` updates the result of the `defaults` query.  
Let me know if this is a suitable solution.

I also added a new test to check if `env.path` actually works.

kdy1 commented:
We don’t apply everything as default

egfx-notifications commented:
> We don’t apply everything as default

Ok, maybe this was phrased wrong. I mean that with the empty query in the autodiscovered files we take the same code path as previously without autodiscovery. So every transformation that was applied to the test cases before the PR without explicit configuration would also be applied now.

kdy1 committed:
Merge branch 'main' into fix/browserslist-config-fallback

kdy1 reviewed (approved):
Thank you! I'll merge it once https://github.com/swc-project/swc/actions/runs/14299826346 is done!

kdy1 reviewed (commented):
~I'll invert this PR with https://github.com/swc-project/swc/pull/10351~

Sorry it was already opt-in.
